### PR TITLE
feat(alerts): 2/4 add additional anomaly detection detectors

### PR DESCRIPTION
**Pull request by andrewm4894**
## Problem

The anomaly detection framework (from #50073, now merged) ships with threshold, z-score, MAD, and ensemble detectors. More algorithms are needed for different data patterns — some handle seasonal data better, others are more robust for complex distributions.

## Changes

https://github.com/user-attachments/assets/69630c64-5c15-418d-83bd-4c2be50c391a

### New detectors

**Statistical detector**:
- IQR (Interquartile Range) — classic box plot method, robust and interpretable

**ML detectors** (via PyOD, already a dependency from PR0):
- COPOD — copula-based, parameter-free
- ECOD — empirical cumulative distribution, parameter-free
- HBOS — histogram-based, very fast for high-volume alerting
- Isolation Forest — tree-based anomaly isolation
- KNN — k-nearest neighbors distance
- LOF — local outlier factor, density-based, good for seasonal data
- OCSVM — one-class SVM, learns boundary around normal data
- PCA — detects via reconstruction error

### Frontend

- Alphabetically sorted detector dropdown with tooltips for all 13 detectors
- Config components with info tooltips on all parameters (sensitivity, neighbors, bins, trees, etc.)
- Multivariate detectors (KNN, LOF, IForest, OCSVM, PCA) default to `diffs_n: 1, lags_n: 3` preprocessing — builds a feature vector from recent deltas instead of running on a single number
- Lag features input shown in Preprocessing section for multivariate detectors only
- `DEFAULT_SINGLE_CONFIGS` is single source of truth for all detector defaults (removed duplicate switch statement)
- Alert `calculation_interval` now defaults to match the insight's interval (hourly insight → hourly alert)
- All detectors use normalized [0,1] probability threshold (replacing stale `contamination` param)

### Backend

- Fixed ensemble `detect()` to properly combine `triggered_indices` (union for OR, intersection for AND) and `all_scores` (element-wise max for OR, min for AND) — was incorrectly using only the last sub-detector's values
- Fixed mypy error: guard `alert.config` for None before using `in` operator
- All detectors registered via `DetectorType` enum consistently
- Added `DETECTOR_MIN_SAMPLES` entries for all new detector types

## How did you test this code?

- 105 unit tests pass: `pytest posthog/tasks/alerts/detectors/test/test_detectors.py -x -q`
- Local end-to-end testing with hourly `test_metric` insight (72h baseline ~50/hr + spike hours):
  - Z-Score: score=0.9999, correctly triggered on spike
  - KNN (5 neighbors, diffs+5 lags): score=0.9923, correctly triggered
  - Isolation Forest (100 trees, diffs+3 lags): score=1.0, correctly triggered
  - Ensemble (Z-Score OR PCA): score=0.9999, correctly triggered with proper union of indices
- OpenAPI types regenerated with `hogli build:openapi`

## Publish to changelog?

No — gated behind `alerts-anomaly-detection` feature flag.

## Docs update

No docs update needed.

## 🤖 LLM context

Co-authored with Claude Code. Key areas of LLM contribution: ensemble combining logic fix, detector UI improvements (sorting, tooltips, lag defaults), alert interval defaulting, and test coverage.

### CODE REVIEW

File: posthog/tasks/alerts/checks.py [hunk 1]
```diff
@@ -345,8 +345,14 @@ def check_alert_and_notify_atomically(alert: AlertConfiguration, capture_ph_even
         # we won't retry and set alert to errored state
         error = {"message": str(err), "traceback": traceback.format_exc()}
 
-    # 2. Check alert value against threshold
-    alert_check = add_alert_check(alert, value, breaches, error)
+    # 2. Extract detector fields and create alert check
+    anomaly_scores = getattr(alert_evaluation_result, "anomaly_scores", None) if alert_evaluation_result else None
+    triggered_points = getattr(alert_evaluation_result, "triggered_points", None) if alert_evaluation_result else None
+    triggered_dates = getattr(alert_evaluation_result, "triggered_dates", None) if alert_evaluation_result else None
+    interval = getattr(alert_evaluation_result, "interval", None) if alert_evaluation_result else None
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/checks.py:373]: Fixed in ea4187cc - initialized `alert_evaluation_result = None` before the try block.

File: posthog/tasks/alerts/trends.py [hunk 2]
```diff
@@ -429,3 +432,125 @@ def _breach_messages(
...
+        else TrendsAlertConfig(type="TrendsAlertConfig", series_index=0)
+    )
+    detector_type = DetectorType(detector_config.get("type", "zscore"))
+
+    # Calculate minimum samples needed for this detector
+    min_samples = DETECTOR_MIN_SAMPLES.get(detector_type, 31)
+
+    # For statistical detectors, check if they have a window config
+    window = detector_config.get("window", 30)
+    if detector_type == DetectorType.ZSCORE:
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/trends.py]: Fixed in ea4187cc - now using `window + 1` for all windowed detectors (Z-Score, MAD, IQR), not just Z-Score.

File: posthog/tasks/alerts/detectors/pyod_detectors/ecod.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        contamination = self.config.get("contamination", 0.1)
+
+        # Reshape for sklearn if 1D
+        if data.ndim == 1:
+            data = data.reshape(-1, 1)
+
+        # Train on all data
+        model = ECOD(contamination=contamination)
+        model.fit(data)
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/pyod_detectors/ecod.py]: This is by design for unsupervised anomaly detection on short time series. PyOD's `predict_proba()` uses a fitted model to score all points — fitting on the full window (including the last point) is the standard approach for transductive anomaly detection. With our typical 30-200 data points, the last point has negligible influence on model parameters.

The statistical detectors (Z-Score, MAD, IQR) use a different approach because their math naturally separates the window from the test point. For PyOD detectors, the fit/score pattern is how the library is designed to be used.

If we see false negatives in practice (anomalies being suppressed by self-influence), we can switch to a leave-one-out scheme, but this hasn't been an issue in testing.

File: posthog/tasks/alerts/detectors/preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    Args:
+        data: Input time series
+        alpha: Smoothing factor (0 < alpha <= 1). Higher = less smoothing.
+    """
+    result = np.zeros_like(data, dtype=float)
+    result[0] = data[0]
+    for i in range(1, len(data)):
+        result[i] = alpha * data[i] + (1 - alpha) * result[i - 1]
+    return result
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/preprocessing.py:78]: This code is from PR0 which has been merged to master — no longer in this PR's diff. The `exponential_smoothing` function was kept as a utility for future preprocessing options but I agree it's dead code. Will clean up in a follow-up if it's still there.

File: posthog/tasks/alerts/detectors/registry.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    from posthog.tasks.alerts.detectors import threshold  # noqa: F401
+    from posthog.tasks.alerts.detectors.statistical import iqr, mad, zscore  # noqa: F401
+
+    # PyOD detectors (optional - require pyod package)
+    try:
+        from posthog.tasks.alerts.detectors.pyod_detectors import copod, ecod, isolation_forest, knn  # noqa: F401
+    except ImportError:
+        logger.warning("PyOD not installed - ML-based detectors (ECOD, COPOD, IsolationForest, KNN) unavailable")
+
+    _REGISTERED = True
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/registry.py]: This is incorrect — the ensemble detector IS implemented and registered. `_ensure_registered()` imports `ensemble` at line 68: `from posthog.tasks.alerts.detectors import ensemble, threshold`. The ensemble is registered via `@register_detector("ensemble")` in `ensemble.py`. It's intentionally NOT in the `DetectorType` enum (only single detectors are) — the registry handles it by string key.

This code is from PR0 which has been merged to master.

File: posthog/tasks/alerts/detectors/statistical/zscore.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            )
+
+        z_score = abs((current_value - mean) / std)
+
+        return DetectionResult(
+            is_anomaly=z_score > threshold,
+            score=float(z_score),
+            triggered_indices=[len(values) - 1] if z_score > threshold else [],
+            all_scores=[float(z_score)],
+            metadata={"mean": float(mean), "std": float(std), "value": float(current_value)},
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/statistical/zscore.py]: This is an intentional design choice from PR0 (now merged to master). `detect()` checks the latest point only and returns a single score — this is the primary use case for alert checking. `detect_batch()` scores all points for backfill/preview purposes (PR2). The calling code in `trends.py` uses `detect()` for alert checks and stores `all_scores` for the check history, where a single score is correct.

No longer in this PR's diff.

File: posthog/tasks/alerts/trends.py [hunk 7]
```diff
@@ -96,6 +101,9 @@ def check_trends_alert(alert: AlertConfiguration, insight: Insight, query: Trend
 
     match condition.type:
         case AlertConditionType.ABSOLUTE_VALUE:
+            if threshold.type != InsightThresholdType.ABSOLUTE:
+                raise ValueError(f"Absolute threshold not configured for alert condition ABSOLUTE_VALUE")
+
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/trends.py:107]: This code is from PR0 which has been merged to master — no longer in this PR's diff. The `threshold` variable is guarded earlier in the function: when `alert.threshold` is None, the code takes the detector-based path (lines 83-85 check `alert.detector_config`), not the threshold-based path. The code at line 105 is only reached when `threshold` is not None.

File: posthog/tasks/alerts/trends.py [hunk 8]
```diff
@@ -120,6 +128,12 @@ def check_trends_alert(alert: AlertConfiguration, insight: Insight, query: Trend
             ):
                 return no_result_evaluation
 
+            if check_current_interval and threshold.bounds.upper is None:
+                # checking for value > X so we can also check current interval value
+                raise ValueError(
+                    f"check_ongoing_interval is only supported for alert condition ABSOLUTE_VALUE when upper threshold is specified"
+                )
+
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/trends.py:138]: Same as above — this code is from PR0 (merged to master). The `threshold` is guaranteed non-None at this point due to the guard earlier in the function.

File: products/alerts/frontend/generated/api.schemas.ts [hunk 9]
```diff
@@ -264,16 +261,108 @@ export interface EnsembleDetectorConfigApi {
...
+=======
+    | ZScoreDetectorConfigApi
+    | MADDetectorConfigApi
+    | IQRDetectorConfigApi
+    | ThresholdDetectorConfigApi
+    | ECODDetectorConfigApi
+    | COPODDetectorConfigApi
+    | IsolationForestDetectorConfigApi
+    | KNNDetectorConfigApi
+>>>>>>> 9b51198b0cb (chore: update OpenAPI generated types)
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/alerts/frontend/generated/api.schemas.ts]: Fixed in 14857c9 — resolved the merge conflict markers and included both ensemble types and the new detector types (ECOD, COPOD, IsolationForest, KNN). Also updated `contamination` fields to `threshold` to match the normalized [0,1] probability score design.

File: frontend/src/lib/components/Alerts/views/DetectorSelector.tsx [hunk 10]
```diff
@@ -267,6 +408,269 @@ function SensitivityInput({
...
+            <SensitivityInput
+                value={config.threshold ?? 0.9}
+                onChange={(val) => onChange({ ...config, threshold: val })}
+                tooltip="Anomaly probability threshold (0-1). Higher = fewer alerts."
+            />
+            <div>
+                <Label text="Number of trees" tooltip="More trees = more accurate but slower. 100 is a good default." />
+                <LemonInput
+                    type="number"
+                    min={10}
```
Comment (andyzzhao) [inline frontend/src/lib/components/Alerts/views/DetectorSelector.tsx:497]: Could we have these constraints on the model too?

File: posthog/tasks/alerts/detectors/pyod_detectors/base_pyod.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+        return DetectionResult(
+            is_anomaly=is_anomaly,
+            score=prob,
+            triggered_indices=[len(data) - 1] if is_anomaly else [],
+            all_scores=[prob],
+            metadata={"raw_score": float(model.decision_function(last_point)[0])},
+        )
+
+    def detect_batch(self, data: np.ndarray) -> DetectionResult:
```
Comment (andyzzhao) [inline posthog/tasks/alerts/detectors/pyod_detectors/base_pyod.py:58]: is there a reason detect_batch doesn't split train and test like detect does?

File: frontend/src/lib/components/Alerts/views/DetectorSelector.tsx (hunk 10 above)
Comment (andrewm4894) [reply frontend/src/lib/components/Alerts/views/DetectorSelector.tsx:497]: Good call — added backend validation in `AlertSerializer._validate_detector_params()` that enforces the same ranges as the frontend:

- `threshold`: 0.0–1.0
- `n_estimators`: 10–500
- `n_neighbors`: 1–50
- `n_bins`: 5–50
- `multiplier`: 0.5–10.0
- `window`: 5–100
- preprocessing `smooth_n`: 0–30, `lags_n`: 0–10

Also validates sub-detector params within ensemble configs.

File: posthog/tasks/alerts/detectors/pyod_detectors/base_pyod.py (hunk 11 above)
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/pyod_detectors/base_pyod.py:58]: Good question — it didn't before, but it does now! Updated `detect_batch` to use a rolling fit: for each point `i`, it fits on `data[:i-offset]` and scores point `i`, exactly mirroring what `detect()` would have done at each timestep.

This gives accurate backfill scores that reflect real-time behavior. The cost is N separate fit calls instead of one, but with our data sizes (30–200 points) and model complexity (PyOD on 1–4 features) each fit is <1ms so the total is negligible.

File: posthog/tasks/alerts/detectors/pyod_detectors/ecod.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/tasks/alerts/detectors/pyod_detectors/ecod.py]: Fixed — refactored all PyOD detectors to use a `BasePyODDetector` base class that:

1. **`detect()`**: fits on `data[:-training_offset_n]`, scores `data[-training_offset_n:]` (default offset=1)
2. **`detect_batch()`**: rolling fit — for each point `i`, fits on `data[:i-offset]` and scores point `i`, exactly matching real-time behavior

The `training_offset_n` config param (default 1) controls how many recent points are excluded from training, which also gives us a knob to control model drift speed in the future.

Each PyOD detector is now ~20 lines, only implementing `_build_model()`.

File: frontend/src/lib/components/Alerts/views/DetectorSelector.tsx (hunk 10 above)
Comment (andrewm4894) [reply frontend/src/lib/components/Alerts/views/DetectorSelector.tsx:497]: Window of 100 max might be a bit small as is just 100 hours so prob could be much more ideally 7 day hourly data as smallest max window for example

**Review by andyzzhao — APPROVED** [review]
